// rank.proto
syntax = "proto3";

package rank;

// 定义服务
service RankService {
  rpc CreateRank (CreateRankRequest) returns (RankResponse);
  rpc GetRank (GetRankRequest) returns (RankResponse);
  rpc UpdateRank (UpdateRankRequest) returns (RankResponse);
  rpc UpdateSendForwardTimes (UpdateSendForwardTimesRequest) returns (RankResponse); // 新增方法
  rpc UpdateSendBackwardTimes (UpdateSendBackwardTimesRequest) returns (RankResponse); // 新增方法
  rpc DeleteRank (DeleteRankRequest) returns (DeleteRankResponse);
  rpc GetRankByStageId (GetRankByStageIdRequest) returns (RankResponse); // 新增方法
  rpc GetSendForwardTimesByStageId (GetSendForwardTimesByStageIdRequest) returns (GetSendForwardTimesResponse); // 新增方法
  rpc GetSendBackwardTimesByStageId (GetSendBackwardTimesByStageIdRequest) returns (GetSendBackwardTimesResponse); // 新增方法
}

// 定义消息
message Rank {
  string id = 1;
  int32 stage_id = 2;
  int32 chunk_id = 3;
  int32 local_rank = 4;
  int32 global_rank = 5;
  int32 sendforwardtimes = 6;
  int32 sendbackwardtimes = 7;
  map<string, int32> latest_getrank_info = 8;  // 使用 map 表示字典
}

message CreateRankRequest {
  int32 stage_id = 1;
  int32 chunk_id = 2;
  int32 local_rank = 3;
  int32 global_rank = 4;
  int32 sendforwardtimes = 5;
  int32 sendbackwardtimes = 6;
  map<string, int32> latest_getrank_info = 7;  // 使用 map 表示字典
}

message GetRankRequest {
  string id = 1;
}

message UpdateRankRequest {
  string id = 1;
  int32 stage_id = 2;
  int32 chunk_id = 3;
  int32 local_rank = 4;
  int32 global_rank = 5;
  int32 sendforwardtimes = 6;
  int32 sendbackwardtimes = 7;
  map<string, int32> latest_getrank_info = 8;
}

message UpdateSendForwardTimesRequest { // 新增消息
  string id = 1;
  int32 sendforwardtimes = 2;
}

message UpdateSendBackwardTimesRequest { // 新增消息
  string id = 1;
  int32 sendbackwardtimes = 2;
}

message DeleteRankRequest {
  string id = 1;
}

message GetRankByStageIdRequest {
  int32 stage_id = 1;
}

message GetSendForwardTimesByStageIdRequest { // 新增消息
  int32 stage_id = 1;
}

message GetSendBackwardTimesByStageIdRequest { // 新增消息
  int32 stage_id = 1;
}

message GetSendForwardTimesResponse { // 新增消息
  int32 sendforwardtimes = 1;
}

message GetSendBackwardTimesResponse { // 新增消息
  int32 sendbackwardtimes = 1;
}

message RankResponse {
  Rank rank = 1;
}

message DeleteRankResponse {
  bool success = 1;
}